pkg/controller/studyjob/utils.go

```diff
@@ -237,7 +251,7 @@ func validateNASJob(instance *katibv1alpha1.StudyJob) error {
 		return fmt.Errorf("Missing OutputSize in NasConfig.GraphConfig: %v", instance.Spec.NasConfig.GraphConfig)
 	}
 
-	if instance.Spec.NasConfig.GraphConfig.NumLayers == 0 && instance.Spec.SuggestionSpec.SuggestionAlgorithm == suggestionAlgorithmRL {
+	if instance.Spec.NasConfig.GraphConfig.NumLayers == 0 && contains(suggestionNASList, instance.Spec.SuggestionSpec.SuggestionAlgorithm) {
```

Comment:
I think line 298 can use contains(suggestionNASList,..), but here I don't think so.
If we add another nas suggestion, it may not need NumLayers.

Reply:
You think we should specify this block only for ```suggestionNASRL``` ?

Reply:
Is there a better place to put algorithm specific checks other than the controller?  This list can grow over time with more algorithms and these checks are very specific to the algorithm implementation. @andreyvelich @hougangliu  Thoughts?

Reply:
From my point of view, it is important to check it in the controller. We don't need to run getSuggestion if something wrong with yaml file. We can't set failed condition for the StudyJob inside Suggestion. Controller do that.

Reply:
We should do algorithm-specific check in suggestion service.

* Add ValidateSuggestionParameter API to Suggestion service 
or
* GetSuggesiton returns a specific error code if the config is invalid then the controller handle the error code

@andreyvelich @johnugeorge @hougangliu Thoughts?

Reply:
> Add ValidateSuggestionParameter API to Suggestion service
Good idea

Reply:
@YujiOshima How it will work, if we add ValidateSuggestionParameter API in Suggestion Service? Can you explain, please? Right now, suggestion return just trials that need to be created by controller.

Reply:
@andreyvelich At initialization, the controller calls ValidateSuggestionParameter and send the parameters of a suggestion algorithm to a corresponding suggestion service.
Then the SuggestionService validates the parameters and if it returns an error, the controller won't start the optimization loop.

Reply:
Do you think we should do it for each Suggestion or make it just for parameters specific for current suggestion?
For example for RL Suggestion parameter Number of Layers here: 
https://github.com/kubeflow/katib/blob/76f8397980979ae2c97932330196933e23d4cfd4/pkg/controller/studyjob/utils.go#L34
and Step here:
https://github.com/kubeflow/katib/blob/76f8397980979ae2c97932330196933e23d4cfd4/pkg/controller/studyjob/utils.go#L298 ?
Other parameters specific to each NAS or HP algorithm, we can validate inside validation function in controller.

Reply:
We should do it for every Suggestion algos.
For example in BO, [this function](https://github.com/kubeflow/katib/blob/76f8397980979ae2c97932330196933e23d4cfd4/pkg/suggestion/bayesian_service.py#L138) should be done in ValidateSuggestionParameter.

And same as GetSuggestion, it can automatically select algorithm which user requested. https://github.com/kubeflow/katib/blob/master/cmd/manager/main.go#L104

Reply:
Should we also put inside Suggestion validation of these parameters: https://github.com/kubeflow/katib/blob/76f8397980979ae2c97932330196933e23d4cfd4/pkg/controller/studyjob/utils.go#L277 ?

Reply:
@YujiOshima In such a design, what should be the validation response  when the suggestion service doesn't exist in the cluster(not yet deployed)? 
Sample scenario: User applies a valid Studyjob config. However, the corresponding Suggestion service doesn't exist in the cluster  (may not be deployed at that instant)

Reply:
@andreyvelich  Both is OK that calling SetSuggestionParameters before validation or put parameters in Validate args.

@johnugeorge 
Same as GetSuggestion, the manager will receive the request like a proxy. If the Suggestion service does not exist or not responsible, the manager can handle and return Unavailable  error code https://github.com/grpc/grpc-go/blob/master/codes/codes.go#L139

Reply:
@YujiOshima In that case, should we send StudyConfig to ValidateSuggestionParameters, because we don't save StudyConfig in db, if it is not valid. 
We can't GetStudy like here: https://github.com/kubeflow/katib/blob/master/pkg/suggestion/nasrl_service.py#L219 ?
As well, should we put this function after this line: https://github.com/kubeflow/katib/blob/master/pkg/controller/studyjob/katib_api_util.go#L43 ?
We generate StudyConfig and can send it to Suggestion to check, whether it is valid or not.

Reply:
@andreyvelich You're right. StudyConf and SuggestionParams should be included in an args of ValidateSuggestionParameters.

Reply:
@YujiOshima  Ok.  In this PR, should I make a validation only for NAS RL service, or we should include BO here, as well?

Reply:
@andreyvelich Only for NAS RL is enough here. If you could add validation for BO in another PR, it would be fantastic.

Reply:
@YujiOshima @hougangliu What do you think about name of this API: ```ValidateStudyParameters``` ? 
It looks confused with ```ValidateSuggestionParameters``` since we validate StudyConfig and SuggestionParameters in this function